drv/ekf/ekf.c

```diff
@@ -0,0 +1,481 @@
+    ekf_p->W += K[4] * y;
+
+    // Constraint theta angle to [-pi, +pi]
+    ekf_p->theta = _EKF_angle_overflow(ekf_p->theta);
+
+    // ##################################################
+    //              Update Posterior P
+    // ##################################################
+    _EKF_update_post_P_gyro_W(ekf_p, K);
+
+}
+
+void EKF_set_state(KalmanFilter_t *ekf_p,float *x0)
+{
+
+    // Copy inital state x0
+    ekf_p->x     = x0[0];
+    ekf_p->y     = x0[1];
+    ekf_p->theta = x0[2];
+    ekf_p->V     = x0[3];
+    ekf_p->W     = x0[4];
+}
+
+void EKF_set_P(KalmanFilter_t *ekf_p,float *P0)
+{
+
+    // Copy the initial Covariance
+    memcpy(ekf_p->P, P0, 25*4); // 25*4 becasue we are copying 25 floats.
+} 
+
+//=========================== private ==========================================
+
+void _EKF_state_transition(KalmanFilter_t *ekf_p,float dt)
+{
+    // Get local copies of the state variables
+    float x = ekf_p->x;
+    float y = ekf_p->y;
+    float theta = ekf_p->theta;
+    float V = ekf_p->V;
+    float W = ekf_p->W;
+
+    // Run the state update
+    ekf_p->x = x + dt * V * cos(theta);
+    ekf_p->y = y + dt * V * sin(theta);
+    ekf_p->theta = theta + dt * W;
+
+    // Regularize the value of Theta between -pi/2 and pi/2
+    ekf_p->theta = _EKF_angle_overflow(ekf_p->theta);
+}
+
+void _EKF_compute_K_lh2_xy(KalmanFilter_t *ekf_p, float *K)
+{
+    // Create a local pointer of P, for clarity (I don't want to have to write ekf_p->P seven million times)
+    float *P = ekf_p->P;
+
+    // Bring a few important values from the R matrix
+    float r00 = EKF_R_LH2__0_0;
+    float r11 = EKF_R_LH2__1_1;
+
+    // Calculate K
+    float denom = (r00 * r11 + r00 * P[6] + r11 * P[0] + P[0] * P[6] - P[1] * P[5]);  // Commonly appearing term
+
+    K[0] = ((r11 + P[6]) * P[0] - P[1] * P[5]) / denom;
+    K[1] = r00 * P[1] / denom;
+    K[2] = r11 * P[5] / denom;
+    K[3] = ((r00 + P[0]) * P[6] - P[1] * P[5]) / denom;
+    K[4] = ((r11 + P[6]) * P[10] - P[5] * P[11]) / denom;
+    K[5] = ((r00 + P[0]) * P[11] - P[1] * P[10]) / denom;
+    K[6] = ((r11 + P[6]) * P[15] - P[5] * P[16]) / denom;
+    K[7] = ((r00 + P[0]) * P[16] - P[1] * P[15]) / denom;
+    K[8] = ((r11 + P[6]) * P[20] - P[5] * P[21]) / denom;
+    K[9] = ((r00 + P[0]) * P[21] - P[1] * P[20]) / denom;
+}
+
+void _EKF_update_post_P_lh2_xy(KalmanFilter_t *ekf_p, float *K)
+{
+
+    // Create a local copy of P. This will help update the actual value of ekf_p->P later
+    float P[25];
+    memcpy(P, ekf_p->P, 25 * 4); // 25*4 becasue we are copying 25 floats.
+
+    // Get values from the R, Noise Matrices.
+    //float r00 = EKF_R_LH2__0_0;
+    //float r11 = EKF_R_LH2__1_1;
+
+    // Update the posterior  Covariance
+    // Implement the following formula
+    //
+    // P_post = (I - K @ H) @ P
+    //
+
+    // Get repeated operations in a variable
+    float temp0 = (1 - K[0]);
+    float temp1 = (1 - K[3]);
+
+    ekf_p->P[0] = temp0 * P[0] - K[1] * P[5];
```

Comment:
Since `P` is the same address as `ekf_p->P`. Why not do:
```suggestion
    P[0] = temp0 * P[0] - K[1] * P[5];
```
?

That would follow the comment above "I don't want to have to write ekf_p->P seven million times"